packages/core/core-flows/src/promotion/steps/add-or-remove-campaign-promotions.ts

```diff
@@ -0,0 +1,64 @@
+import { ModuleRegistrationName } from "@medusajs/modules-sdk"
+import { IPromotionModuleService, LinkWorkflowInput } from "@medusajs/types"
+import { StepResponse, WorkflowData, createStep } from "@medusajs/workflows-sdk"
+
+export const addOrRemoveCampaignPromotionsStepId =
+  "add-or-remove-campaign-promotions"
+export const addOrRemoveCampaignPromotionsStep = createStep(
+  addOrRemoveCampaignPromotionsStepId,
+  async (input: WorkflowData<LinkWorkflowInput>, { container }) => {
+    const {
+      id: campaignId,
+      add: promotionIdsToAdd = [],
+      remove: promotionIdsToRemove = [],
+    } = input
+    const promotionModule = container.resolve<IPromotionModuleService>(
+      ModuleRegistrationName.PROMOTION
+    )
+
+    if (promotionIdsToAdd.length) {
+      await promotionModule.addPromotionsToCampaign({
+        id: campaignId,
+        promotion_ids: promotionIdsToAdd,
+      })
+    }
+
+    if (promotionIdsToRemove.length) {
+      await promotionModule.removePromotionsFromCampaign({
```

Comment:
`addPromotionsToCampaign` and `removePromotionsFromCampaign` have to be two different steps as the DB transaction is not shared.

Reply:
good catch!